src/gwt/src/org/rstudio/studio/client/workbench/views/source/NewShinyWebApplication.java

```diff
@@ -0,0 +1,172 @@
+{
+   public static class Result
+   {
+      public Result(String appName, String appType, String appDir)
+      {
+         appName_ = appName;
+         appType_ = appType;
+         appDir_ = appDir;
+      }
+      
+      public String getAppName() { return appName_; }
+      public String getAppType() { return appType_; }
+      public String getAppDir() { return appDir_; }
+      
+      private String appName_;
+      private String appType_;
+      private String appDir_;
+   }
+   
+   private static class VerticalSpacer extends Composite
+   {
+      public VerticalSpacer(String height)
+      {
+         FlowPanel panel = new FlowPanel();
+         panel.setHeight(height);
+         initWidget(panel);
+      }
+   }
+   
+   public NewShinyWebApplication(String caption, 
+                                 String workingDirectory,
+                                 OperationWithInput<Result> operation)
+   {
+      super(caption, operation);
+      
+      container_ = new VerticalPanel();
+      
+      // Create individual widgets
+      appNameLabel_ = new Label("Application Name:");
+      appNameLabel_.addStyleName(RES.styles().label());
+      
+      appNameTextBox_ = new TextBox();
+      appNameTextBox_.getElement().getStyle().setMarginLeft(10, Unit.PX);
+      appNameTextBox_.getElement().setAttribute("placeholder", "Name");
+      appNameTextBox_.setText(defaultAppName(workingDirectory));
+      
+      appTypeLabel_ = new Label("Application Type:");
+      appTypeLabel_.addStyleName(RES.styles().label());
+      
+      appTypeSelectWidget_ = new SelectWidget(
+            "",
+            new String[] {
+                  "Single File (app.R)",
+                  "Multiple File (ui.R/server.R)",
+            },
+            new String[] {
+                  TYPE_SINGLE_FILE,
+                  TYPE_MULTI_FILE
+            },
+            false,
+            true,
+            false);
+      
+      appTypeSelectWidget_.getElement().getStyle().setMarginTop(0, Unit.PX);
+      appTypeSelectWidget_.getElement().getStyle().setMarginBottom(0, Unit.PX);
+      
+      directoryChooserTextBox_ = new DirectoryChooserTextBox("Directory:", null);
+      directoryChooserTextBox_.setText(workingDirectory);
+      
+      // Add them to parent
+      Grid appNameTypeGrid = new Grid(3, 2);
+      appNameTypeGrid.setWidget(0, 0, appNameLabel_);
+      appNameTypeGrid.setWidget(0, 1, appNameTextBox_);
+      
+      appNameTypeGrid.setWidget(1, 0, new VerticalSpacer("4px"));
+      appNameTypeGrid.setWidget(1, 1, new VerticalSpacer("4px"));
+      
+      appNameTypeGrid.setWidget(2, 0, appTypeLabel_);
+      appNameTypeGrid.setWidget(2, 1, appTypeSelectWidget_);
+      container_.add(appNameTypeGrid);
+      
+      container_.add(new VerticalSpacer("12px"));
+      container_.add(directoryChooserTextBox_);
+      
+      Scheduler.get().scheduleDeferred(new ScheduledCommand()
```

Comment:
You can override onDialogShown for this: https://github.com/rstudio/rstudio/blob/master/src/gwt/src/org/rstudio/core/client/widget/ModalDialogBase.java#L194-L196